[Feature][Hive JDBC Source] Support Hive JDBC Source Connector

NickCodeJourney opened the pull request:
close #5389
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

## Purpose of this pull request

<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [x] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

Commit:
:sparkles: "Added Hive support to the JDBC connector

This commit includes the addition of Hive JDBC dependencies in the pom.xml file to enable the JDBC connector support Hive. The Hive dialect, type mapper and row converter were added accordingly. This update allows users to connect and interact with Hive directly using the JDBC connector."

Commit:
:sparkles: "Added Hive support to the JDBC connector

This commit includes the addition of Hive JDBC dependencies in the pom.xml file to enable the JDBC connector support Hive. The Hive dialect, type mapper and row converter were added accordingly. This update allows users to connect and interact with Hive directly using the JDBC connector."

Commit:
:art: Refactor HiveTypeMapper for JDBC connector

zhilinli123 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveTypeMapper.java:
```diff
@@ -0,0 +1,120 @@
+            case HIVE_DOUBLE_PRECISION:
+                return BasicType.DOUBLE_TYPE;
+            case HIVE_DECIMAL:
+            case HIVE_NUMERIC:
+                if (precision > 0) {
+                    return new DecimalType(precision, metadata.getScale(colIndex));
+                }
+                LOG.warn("decimal did define precision,scale, will be Decimal(38,18)");
+                return new DecimalType(38, 18);
```
```
                if (scale == 0) {
                    if (precision <= 9) {
                        return BasicType.INT_TYPE;
                    } else if (precision <= 18) {
                        return BasicType.LONG_TYPE;
                    } else if (precision <= 38) {
                        return new DecimalType(38, 0);
                    }
                }
                return new DecimalType(38, 18);
```

NickCodeJourney replied:
I think it should maintain similar data types

EricJoy2048 reviewed (commented):
Please add document for this connector reference https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/Mysql.md

EricJoy2048 commented:
Please add e2e for this connector reference https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java

Commit:
:sparkles: Add Hive JDBC source connector documentation

This commit introduces a new documentation file under `/docs/en/connector-v2/source/` explaining how to use the Hive JDBC source connector. The included information covers the supported engines, key features, support DataSource info, data type mapping, source options and database dependency, along with examples for how to use the connector with different configurations. This documentation is intended to help users to understand and effectively use the Hive JDBC source connector in their data pipeline.

Commit:
:sparkles: Add JDBC Hive IT and source config to SeaTunnel project

Introduces JDBC Hive Integration Tests and new configuration settings for the 'connector-jdbc-e2e' module in the SeaTunnel v2 project. These additions are designed to enhance testing and configuration handling capabilities for connections to Hive databases. The new JdbcHiveIT class implements Hive-specific configurations and integration test cases. The jdbc_hive_source.conf file contains a sample configuration setup for a Hive JDBC source connection.

Commit:
:fire: Refactored JDBC configuration filename in e2e-test

Fixed the case sensitivity issue in jdbc config filename in e2e-test. Changed "/jdbc_HIVE_source_and_sink.conf" to "/jdbc_hive_source_and_sink.conf". Linux servers are case sensitive hence the filename should match exact case to avoid file not found exceptions while running end-to-end testing.

EricJoy2048 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveTypeMapper.java:
```diff
@@ -0,0 +1,120 @@
+        String columnType = metadata.getColumnTypeName(colIndex).toUpperCase();
+        int precision = metadata.getPrecision(colIndex);
+        switch (columnType) {
+            case HIVE_TINYINT:
```
Hive `HIVE_TINYINT` need map to seatunnel type `BYTE_TYPE` ?

EricJoy2048 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveTypeMapper.java:
```diff
@@ -0,0 +1,120 @@
+                return BasicType.STRING_TYPE;
+            case HIVE_BOOLEAN:
+                return BasicType.BOOLEAN_TYPE;
+            case HIVE_BINARY:
```
How about map `HIVE_BINARY ` to  `PrimitiveByteArrayType.INSTANCE`?

NickCodeJourney replied:
don't support 
`Caused by: java.sql.SQLFeatureNotSupportedException: Method not supported
	at org.apache.hive.jdbc.HiveBaseResultSet.getBytes(HiveBaseResultSet.java:222)
	at org.apache.seatunnel.connectors.seatunnel.jdbc.internal.converter.AbstractJdbcRowConverter.toInternal(AbstractJdbcRowConverter.java:97)
	at org.apache.seatunnel.connectors.seatunnel.jdbc.internal.JdbcInputFormat.nextRecord(JdbcInputFormat.java:220)
	... 6 more`

`CREATE TABLE all_data_types
(
    int_column              INT,
    integer_column          INTEGER,
    bigint_column           BIGINT,
    smallint_column         SMALLINT,
    tinyint_column          TINYINT,
    double_column           DOUBLE,
    double_PRECISION_column DOUBLE PRECISION,
    float_column            FLOAT,
    string_column           STRING,
    char_column             CHAR(10),
    varchar_column          VARCHAR(20),
    boolean_column          BOOLEAN,
    date_column             DATE,
    timestamp_column        TIMESTAMP,
    binary_column           BINARY,
    decimal_column          DECIMAL(10, 2),
    numeric_column          NUMERIC(10, 2)
);`
`INSERT INTO all_data_types
VALUES (2,
        1,
        1234567890,
        32767,
        127,
        123.45,
        123.45,
        67.89,
        'Hello, Hive',
        'CharCol',
        'VarcharCol',
        TRUE,
        '2023-09-04',
        '2023-09-04 10:30:00',
        UNHEX('48656C6C6F2C2048697665'),
        42.12,
        42.12)`


but it can map HIVE_BINARY  to  STRING_TYPE.

EricJoy2048 replied:
Get it.

EricJoy2048 commented on seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcHiveIT.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.Map;
+
+@Slf4j
+public class JdbcHiveIT extends AbstractJdbcIT {
```
I don't found the test case to check the read data from hive is expected. You can reference 
https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-clickhouse-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/ClickhouseIT.java method `compareResult` to check this.

NickCodeJourney replied:
But, I only hava source connector, don't have sink connector. How can I  get sink data to compare with source data

EricJoy2048 replied:
> But, I only hava source connector, don't have sink connector. How can I get sink data to compare with source data

You can use `AssertSink` to check data. Please reference `seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/resources/sql_transform/func_system.conf`

NickCodeJourney replied:
AssertSink has some issues and cannot test the decimal type.

<img width="1131" alt="image" src="https://github.com/apache/seatunnel/assets/72905543/616eed7f-3ede-4ca9-bdaa-f18f4a496ee6">
<img width="550" alt="image" src="https://github.com/apache/seatunnel/assets/72905543/60649b2f-333c-4ace-8cdc-f825ec9d2102">

NickCodeJourney replied:
When comparing float and double values, we need to consider precision issues associated with floating-point representation. Converting a float to a double can introduce minor inaccuracies. Even though two values might seem equivalent logically, the IEEE floating-point representation might differ slightly, causing the comparison to fail.

NickCodeJourney replied:
I will pull a request support 'DECIMAL' and fix the bug.

EricJoy2048 replied:
> I will pull a request support 'DECIMAL' and fix the bug.

Thanks, This is exciting.

NickCodeJourney replied:
#5479

EricJoy2048 replied:
> #5479

#5479 was merged.

Commit:
:art: Add BYTE_TYPE mapping in HiveTypeMapper

This commit adds a BYTE_TYPE mapping to the HiveTypeMapper class for the HIVE_TINYINT field. This change allows mapping of HIVE_TINYINT to BasicType.BYTE_TYPE, ensuring type consistency.

zhilinli123 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveTypeMapper.java:
```diff
@@ -0,0 +1,120 @@
+            case HIVE_DOUBLE_PRECISION:
+                return BasicType.DOUBLE_TYPE;
+            case HIVE_DECIMAL:
+            case HIVE_NUMERIC:
+                if (precision > 0) {
+                    return new DecimalType(precision, metadata.getScale(colIndex));
+                }
+                LOG.warn("decimal did define precision,scale, will be Decimal(38,18)");
+                return new DecimalType(38, 18);
```
> I think it should maintain similar data types

You are correct.

zhilinli123 commented on seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/src/test/resources/jdbc_hive_source.conf:
```diff
@@ -0,0 +1,46 @@
+    Jdbc {
+        url = "jdbc:hive2://localhost:10000/default"
+        user = "root"
+        driver = "org.apache.hive.jdbc.HiveDriver"
```
We currently do not have hive E2e tests, can your E2e directly cover jdbc tests and hive thrift mode read and write, metastore
https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/Hive.md,https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/sink/Hive.md

EricJoy2048 replied:
> https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/Hive.md,https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/sink/Hive.md

`jdbc_hive` and `hive` are difference connector, It would be better to separate the e2e of the two connectors. We are missing the e2e for the hive connector. 

@NickYoungPeng  If you are willing, you can provide a separate PR to add the e2e for the hive connector.

Commit:
Merge remote-tracking branch 'origin1/dev' into hiveJdbc

# Conflicts:
#	seatunnel-connectors-v2/connector-jdbc/pom.xml

Commit:
:art: Refactor JDBC connector tests for Hive

Removed redundant file and tidied up pom.xml. Removed the file jdbc_hive_source.conf, added needed dependencies for Hive in pom.xml, and removed an unused module from connector-file pom. Improved tests by updating test configuration to better demonstrate features and modified corresponding test class accordingly. Changes aim to enhance test readability and correctness.

NickCodeJourney commented:
I want Hive JDBC to support Kerberos authentication. Do you have any suggestions?

1. My 1st idea is to enhance the SimpleJdbcConnectionProvider. This would involve modifying the JdbcConnectionConfig class to support Hive JDBC Kerberos authentication.
2. My 2ed idea is Create a new JdbcConnectionProvider, such as HiveJdbcConnectionProvider, and inject it through SPI (Service Provider Interface).

Commit:
:art: Update Hive Image and Add Environment Variable

Updated the HIVE_IMAGE from "youngyangp/hive_3.1.2_arm:1.0.0" to "apache/hive:3.1.3" in the JdbcHiveIT test. Also, introduced "SERVICE_NAME" as an environment variable to the Hive server container.

The HIVE_IMAGE was updated for using an official Apache hive image. The environment variable is crucial for service internal recognition.

Commit:
:art: Refactor styling in JdbcHiveIT file

Uniformly formatted the environment variables for better code readability. The change is on line 160 from '.withEnv("SERVICE_NAME","hiveserver2")' to '.withEnv("SERVICE_NAME", "hiveserver2")'. The whitespace after file.

Commit:
Added auto_commit to jdbc_hive config file

Added a new parameter, 'auto_commit', to the jdbc_hive configuration file to improve database transaction management. This is pivotal for applications that require a high level of data integrity and consistency, or for those that need to manage complex, multi-step transactions.

Commit:
Update hive.jdbc version and remove 'provided' scope

Updated hive.jdbc.version to 3.1.3 in connector-jdbc/pom.xml to have access to the latest features, bug fixes and improvements. Also, removed the 'provided' scope from the same file as it is not necessary and its presence could lead to potential issues with the dependency resolution during the build process.

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

EricJoy2048 commented:
> I want Hive JDBC to support Kerberos authentication. Do you have any suggestions?
> 
> 1. My 1st idea is to enhance the SimpleJdbcConnectionProvider. This would involve modifying the JdbcConnectionConfig class to support Hive JDBC Kerberos authentication.
> 2. My 2ed idea is Create a new JdbcConnectionProvider, such as HiveJdbcConnectionProvider, and inject it through SPI (Service Provider Interface).

It seems that the second option is more suitable.

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

Commit:
:sparkles: "Added kerberos support in JDBC connections

This commit enables kerberos authentication while establishing JDBC connections. The JdbcOptions class has been updated to include new options such as 'use_kerberos', 'kerberos_principal', etc to configure Kerberos settings. A new connection provider 'HiveJdbcConnectionProvider' has been added which takes care of connection creation when using Kerberos. This change also makes the connection provider more flexible as it can now be selected based on the JDBC dialect."

Commit:
:sparkles: "Added kerberos support in JDBC connections

This commit enables kerberos authentication while establishing JDBC connections. The JdbcOptions class has been updated to include new options such as 'use_kerberos', 'kerberos_principal', etc to configure Kerberos settings. A new connection provider 'HiveJdbcConnectionProvider' has been added which takes care of connection creation when using Kerberos. This change also makes the connection provider more flexible as it can now be selected based on the JDBC dialect."

Commit:
Merge remote-tracking branch 'origin/hivejdbc' into hivejdbc-1

Commit:
Add ASF license header to HiveJdbcConnectionProvider file

This commit adds the Apache License Version 2.0 header to the HiveJdbcConnectionProvider file. Rights, terms, and conditions associated with the usage of the file in compliance with ASF policies are now clearly outlined within the header of the file.

Commit:
revert:  "Add Kingbase8 dependency to JDBC connector

The Kingbase8 database connector was added to the JDBC connector's pom.xml. This change allows our application to interact with Kingbase8 databases, expanding possible integration scenarios for applications using our connector."

Commit:
:art: Refactor HiveJdbcConnectionProvider and SimpleJdbcConnectionProvider classes

Changed the method of accessing JdbcConfig object and the Connection object in both classes from getter methods to direct access. This was done to create a more cohesive codebase and promote a uniform way of accessing these objects. This change also aids in improving the code readability and reducing ambiguities when dealing with these objects across different packages.

Commit:
:art: Refactor HiveJdbcConnectionProvider and SimpleJdbcConnectionProvider classes

Changed the method of accessing JdbcConfig object and the Connection object in both classes from getter methods to direct access. This was done to create a more cohesive codebase and promote a uniform way of accessing these objects. This change also aids in improving the code readability and reducing ambiguities when dealing with these objects across different packages.

Commit:
解决冲突

Commit:
解决冲突

Commit:
解决冲突

Commit:
revert

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

Commit:
revert

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

XiaoJiang521 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/config/JdbcOptions.java:
```diff
@@ -161,4 +161,31 @@ public interface JdbcOptions {
+                            "When use kerberos, we should set kerberos principal file path such as '/home/test/test_user.keytab'. ");
+
+    Option<String> KRB5_PATH =
+            Options.key("krb5_path")
```
How about using the name  kerberos_krb5_conf_path so that the kerberos_keytab_path can be equivalent and also be more universal

XiaoJiang521 commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveJdbcConnectionProvider.java:
```diff
@@ -0,0 +1,80 @@
+                        jdbcConfig.kerberosPrincipal, jdbcConfig.kerberosKeytabPath);
+            } catch (IOException e) {
+                throw new JdbcConnectorException(KERBEROS_AUTHENTICATION_FAILED, e);
+            }
```
public class HiveJdbcUtils {

    public static synchronized void doKerberosAuthentication(JdbcConnectionConfig jdbcConfig) {
        String principal = jdbcConfig.kerberosPrincipal;
        String keytabPath = jdbcConfig.kerberosKeytabPath;
        String krb5Path = jdbcConfig.krb5Path;
        System.setProperty("java.security.krb5.conf", krb5Path);
        Configuration configuration = new Configuration();

        if (StringUtils.isBlank(principal) || StringUtils.isBlank(keytabPath)) {
            log.warn(
                    "Principal [{}] or keytabPath [{}] is empty, it will skip kerberos authentication",
                    principal,
                    keytabPath);
        } else {
            configuration.set("hadoop.security.authentication", "kerberos");
            UserGroupInformation.setConfiguration(configuration);
            try {
                log.info(
                        "Start Kerberos authentication using principal {} and keytab {}",
                        principal,
                        keytabPath);
                UserGroupInformation.loginUserFromKeytab(principal, keytabPath);
                log.info("Kerberos authentication successful");
            } catch (IOException e) {
                String errorMsg =
                        String.format(
                                "Kerberos authentication failed using this "
                                        + "principal [%s] and keytab path [%s]",
                                principal, keytabPath);
                throw new JdbcConnectorException(KERBEROS_AUTHENTICATION_FAILED, errorMsg, e);
            }
        }
    }
}
How about pulling out kb authentication this way, so that some logs are added to ensure that multiple sets of kb authentication are used at the same time

XiaoJiang521 replied:
And Later sink supports writing and can also reuse this kb util

Commit:
:art:

Commit:
Merge remote-tracking branch 'origin/hivejdbc' into hivejdbc-1

Commit:
:art: Refactor kerberos authentication for HiveJdbcConnection

EricJoy2048 commented:
Please fix CI error.

Commit:
Added "provided" scope to hive-service dependency in pom.xml

Carl-Zhou-CN commented on seatunnel-connectors-v2/connector-jdbc/pom.xml:
```diff
@@ -150,6 +151,17 @@
+            <dependency>
+                <groupId>org.apache.hive</groupId>
+                <artifactId>hive-jdbc</artifactId>
+                <version>${hive.jdbc.version}</version>
```
${hive.version}

Carl-Zhou-CN commented on seatunnel-connectors-v2/connector-jdbc/pom.xml:
```diff
@@ -47,6 +47,7 @@
         <vertica.version>12.0.3-0</vertica.version>
         <postgis.jdbc.version>2.5.1</postgis.jdbc.version>
         <kingbase8.version>8.6.0</kingbase8.version>
+        <hive.jdbc.version>3.1.3</hive.jdbc.version>
```
<hive.version>3.1.3</hive.version>

Carl-Zhou-CN commented on seatunnel-connectors-v2/connector-jdbc/pom.xml:
```diff
@@ -150,6 +151,17 @@
+            <dependency>
+                <groupId>org.apache.hive</groupId>
+                <artifactId>hive-service</artifactId>
+                <version>${hive.jdbc.version}</version>
```
ditto

Commit:
:sparkles: Add hive-service JAR to JDBC test setup

Commit:
:art: format

Commit:
:art: format

Commit:
:art: format

Commit:
Add Hive dependencies to pom.xml

Commit:
Added Apache Hive JDBC and Hive Service dependencies to the assembly-bin-ci.xml file, as they were missing. This change ensures that Hive-related tasks can be correctly executed without runtime errors due to missing libraries.

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
Update Hive JDBC driver URL in JdbcHiveIT test

Commit:
revert

Commit:
revert

Commit:
revert

Commit:
"Remove unnecessary comment in HiveTypeMapper"

NickCodeJourney committed:
Merge branch 'dev' into hivejdbc

Commit:
:art: format

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

hailin0 reviewed (commented):
Please fix code conflict

NickCodeJourney committed:
Merge branch 'dev' into hivejdbc

Commit:
resolve conflict

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

Commit:
:art: resolve conflict

Commit:
Merge remote-tracking branch 'origin/hivejdbc' into hivejdbc

# Conflicts:
#	seatunnel-connectors-v2/connector-jdbc/pom.xml
#	seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/config/JdbcConnectionConfig.java
#	seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/config/JdbcOptions.java
#	seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveDialect.java
#	seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveJdbcConnectionProvider.java
#	seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/hive/HiveJdbcRowConverter.java
#	seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/pom.xml
#	seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcHiveIT.java

Commit:
add license header

Commit:
:fire:

Commit:
:fire:

Commit:
format

Commit:
Add Hive JDBC dependency and update SimpleJdbcConnectionProvider

This commit introduces the Apache Hive JDBC connector as a dependency in the POM.xml to support connectivity with Hive. It also removes @Data annotation from SimpleJdbcConnectionProvider and manually implements getters and setters. This was done to improve code clarity and ensure explicit control over the modifications to Connection object.

Commit:
Add Hive service as dependency in pom.xml

This commit includes Hive service and Hive JDBC in the dependencies of the pom.xml file under seatunnel-connectors-v2/connector-jdbc. The Hive service has been added to expand the functionality of the connector-jdbc to support more databases, and enhance its capabilities of handling various types of databases.

Commit:
Remove hive-service dependency from pom.xml

This commit removes hive-service from the dependencies of the pom.xml file under seatunnel-connectors-v2/connector-jdbc. It also includes exclusions in the hive-jdbc dependency to prevent unnecessary inclusion of jdk.tools. These changes are done to avoid certain version conflict issues and to optimize the project's dependencies.

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

NickCodeJourney committed:
Merge branch 'dev' into hivejdbc

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

NickCodeJourney committed:
Update dependency in connector-jdbc pom.xml

Updated the dependency to the oceanbase-client from the com.oceanbase group in the connector-jdbc pom.xml file. This change is necessary to maintain API consistency and efficiency.

Carl-Zhou-CN commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java:
```diff
@@ -27,8 +27,9 @@ public enum JdbcConnectorErrorCode implements SeaTunnelErrorCode {
-
+    NO_SUITABLE_DIALECT_FACTORY("JDBC-06", "No suitable dialect factory found"),
+    DONT_SUPPORT_SINK("JDBC-07", "The jdbc type don't support sink"),
+    KERBEROS_AUTHENTICATION_FAILED("JDBC-08", "Kerberos authentication failed");
```
Can you help me update this exception file: https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/Error-Quick-Reference-Manual.md

NickCodeJourney replied:
yes

NickCodeJourney committed:
Add new error codes to Error Quick Reference Manual

Introduced new error codes JDBC-07 and JDBC-08 in connection with the JDBC in the Error Quick Reference Manual. These additions are important to help users understand and diagnose the new issues related to unsupported jdbc type and failed Kerberos authentication respectively.

NickCodeJourney committed:
Update default krb5 path in jdbc configs

Updated the default krb5 file path in jdbc configuration from '/etc/krb5.conf' to '/seatunnel/krb5.conf' or use the default path '/etc/krb5.conf'. This change was made to provide flexibility if the krb5.conf file is not found at the default location. The documentation has been updated to reflect this change

NickCodeJourney committed:
Merge branch 'dev' into hivejdbc

zhilinli123 commented on docs/en/connector-v2/source/Hive-jdbc.md:
```diff
@@ -0,0 +1,160 @@
+
+> JDBC Hive Source Connector
+
+## Support Those Engines
+
+> Spark<br/>
+> Flink<br/>
+> SeaTunnel Zeta<br/>
```
Add Supported Versions

NickCodeJourney replied:
Definitely supports 3.1.3 and 3.1.2, other versions need to be tested

NickCodeJourney committed:
Add Hive version support information to docs

Included definite support information for Hive version 3.1.3 and 3.1.2 in JDBC Hive Source Connector doc, while stating that other versions need testing. This provides clarity on version compatibility, helping users determine if their Hive version is definitely compatible or requires testing.

NickCodeJourney committed:
Merge remote-tracking branch 'origin/hivejdbc' into hiveJdbc

zhilinli123 commented:
fix ci

NickCodeJourney committed:
Merge branch 'apache:dev' into hivejdbc

Carl-Zhou-CN reviewed (approved):
+1

Hisoka-X commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/HiveJdbcUtils.java:
```diff
@@ -0,0 +1,68 @@
+import static org.apache.seatunnel.connectors.seatunnel.jdbc.exception.JdbcConnectorErrorCode.KERBEROS_AUTHENTICATION_FAILED;
+
+@Slf4j
+public class HiveJdbcUtils {
```
ignore this.
~~I found this file never be used, so can we delete it? PS: that's meaning hive jdbc doesn't support kerberos at now.~~
